### PR TITLE
Add DateOnly.ToString / TimeOnly.ToString query mapping for SQL Server

### LINKED ISSUE #31035: Translate DateOnly.ToString() / TimeOnly.ToString() ??
Should it be added to ?:

https://github.com/dotnet/efcore/blob/main/src/EFCore.SqlServer/Query/Internal/SqlServerObjectToStringTranslator.cs

### PR DESCRIPTION
**Pull request by sulton-max**
Summary of changes 

- Added Translator to the list of `SqlServerObjectToStringTranslators`
- Added base test to `GearsOfWarQueryTestBase`
- Added overloads of the base test to `GearsOfWarQuerySqlServerTest`, `TPCGearsOfWarQuerySqlServerTest` and `TPTGearsOfWarQuerySqlServerTest`

- Fixes #31035

### CODE REVIEW

**Comment by sulton-max** [comment]
> 

@dotnet-policy-service agree

File: src/EFCore.SqlServer/Query/Internal/SqlServerDateOnlyMethodTranslator.cs [hunk 1]
```diff
@@ -66,6 +66,18 @@ public SqlServerDateOnlyMethodTranslator(ISqlExpressionFactory sqlExpressionFact
             return _sqlExpressionFactory.Convert(arguments[0], typeof(DateOnly));
         }
 
+
```
Comment (ErikEJ) [inline src/EFCore.SqlServer/Query/Internal/SqlServerDateOnlyMethodTranslator.cs:69]: Dont think you need this, just the one line in the ObjectToStringTranslator should be enough

File: src/EFCore.SqlServer/Query/Internal/SqlServerObjectToStringTranslator.cs [hunk 2]
```diff
@@ -31,6 +31,7 @@ public class SqlServerObjectToStringTranslator : IMethodCallTranslator
             { typeof(decimal), $"varchar({DefaultLength})" },
             { typeof(char), "varchar(1)" },
             { typeof(DateTime), $"varchar({DefaultLength})" },
+            { typeof(DateOnly), $"varchar({DefaultLength})" },
```
Comment (ErikEJ) [inline src/EFCore.SqlServer/Query/Internal/SqlServerObjectToStringTranslator.cs:34]: And TimeOnly? See https://github.com/ErikEJ/EFCore.SqlServer.DateOnlyTimeOnly/commit/e7b8b71994a4aaac92055f1d30f79a45b6b0d310 for my implementation

Comment (sulton-max) [reply src/EFCore.SqlServer/Query/Internal/SqlServerObjectToStringTranslator.cs:34]: Wanted to add with separate commit

File: src/EFCore.SqlServer/Query/Internal/SqlServerDateOnlyMethodTranslator.cs (hunk 1 above)
Comment (sulton-max) [reply src/EFCore.SqlServer/Query/Internal/SqlServerDateOnlyMethodTranslator.cs:69]: But without this test are failing due to different result from query after query rewrite in `QueryAssertions`, should I provide more details on this ?

File: src/EFCore.SqlServer/Query/Internal/SqlServerObjectToStringTranslator.cs (hunk 2 above)
Comment (roji) [reply src/EFCore.SqlServer/Query/Internal/SqlServerObjectToStringTranslator.cs:34]: I think it's better to just add TimeOnly; we'll end up squashing the PR to one commit anyway, and this avoids several back and forths of reviewing rounds.

**Comment by roji** [comment]
@sulton-max please take a look at the failing tests.

**Comment by sulton-max** [comment]
> @sulton-max please take a look at the failing tests.

> @sulton-max please take a look at the failing tests.

Need help on this

Issue - DateOnly standard format 

Currently I'm having 2 issues

1. SQL Lite can only work with "yyyy-mm-dd" format, which is ISO 8601
2. Changing thread culture creates other issues - will address after first one

So
- default value when converting DateOnly to string in c# - 7/20/2023 ( "m/d/yyyy" in my current culture )
- SQL Lite can't work with "yyyy/mm/dd" too, only "yyyy-mm-dd"

I guess because data we are using in`GearsOfWarData` will be converted for DateOnly to "m/d/yyyy" querying it right ?

How can I solve this ? 

Tried multiple formats for SQL Lite in `SqliteDateOnlyMethodTranslator` with following logic using both `strftime` and `date` functions

```
        if (instance?.Type == typeof(DateOnly)
         && method.Name == nameof(DateOnly.ToString))
        {
            return _sqlExpressionFactory.Function(
              "strftime",
              new[] { _sqlExpressionFactory.Constant("%m/%d/%Y"), instance },
              nullable: true,
              argumentsPropagateNullability: new[] { false, true },
              returnType: typeof(string));
        }
```

File: src/EFCore.SqlServer/Query/Internal/SqlServerObjectToStringTranslator.cs (hunk 2 above)
Comment (sulton-max) [reply src/EFCore.SqlServer/Query/Internal/SqlServerObjectToStringTranslator.cs:34]: Ok, will send in this commit, as soon as fixing tests for SQL Lite

File: src/EFCore.SqlServer/Query/Internal/SqlServerDateOnlyMethodTranslator.cs [hunk 3]
```diff
@@ -66,6 +66,18 @@ public SqlServerDateOnlyMethodTranslator(ISqlExpressionFactory sqlExpressionFact
...
+        if (instance?.Type == typeof(DateOnly)
+            && method.Name == nameof(DateOnly.ToString))
+        {
+            return _sqlExpressionFactory.Function(
+                "FORMAT",
+                new[] { instance, _sqlExpressionFactory.Constant("M/d/yyyy") },
```
Comment (ErikEJ) [inline src/EFCore.SqlServer/Query/Internal/SqlServerDateOnlyMethodTranslator.cs:75]: I think the literal format from the type mapping will be used by default, this format will break for sure.

**Comment by roji** [comment]
@sulton-max please do the same thing for DateOnly/TimeOnly that we already do for DateTime - no more and no less; as @ErikEJ wrote above, there shouldn't be a DateOnly-specific translation with FORMAT, which uses a hard-coded United States date format. See the DateTime tests for ToString to understand how it should work.

File: src/EFCore.SqlServer/Query/Internal/SqlServerDateOnlyMethodTranslator.cs (hunk 3 above)
Comment (sulton-max) [reply src/EFCore.SqlServer/Query/Internal/SqlServerDateOnlyMethodTranslator.cs]: Ok, got you

**Comment by sulton-max** [comment]
> @sulton-max please do the same thing for DateOnly/TimeOnly that we already do for DateTime - no more and no less; as @ErikEJ wrote above, there shouldn't be a DateOnly-specific translation with FORMAT, which uses a hard-coded United States date format. See the DateTime tests for ToString to understand how it should work.

Ok, I removed method translators for `DateOnly` and `TimeOnly`, only added them in `SqlServerObjectToStringTranslator` just as `DateTime`

@roji I thought about that initially, but had a hard time finding where are `DateTime` `ToString` tests are, should I search references from `GearsOfWarData` entities ? Found `DateTime` on `CogTag.IssueDate` but it doesn't have any references to `ToString` tests

**Comment by roji** [comment]
See Object_to_string_conversion.

Note that we cannot assert on the actual results from ToString(); since the translation is a SQL cast, every database has its own string representation when a DateTime (or DateOnly) is cast to a string. This is by design.

**Comment by sulton-max** [comment]
@roji 

Oh, I see now. Should I add `DateOnly` / `TimeOnly` tests as part of `Object_to_string_conversion` and delete tests I wrote in `GearsOfWarQueryTestBase` and its overrides ?

So overall there must be 3 tests right ? ( in 3 overridden versions of `Object_to_string_conversion` for SQL Server, SQL Lite and Cosmo DB )

**Comment by roji** [comment]
Yes

File: test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs [hunk 4]
```diff
@@ -2796,7 +2798,7 @@ public virtual Task Comparing_two_collection_navigations_composite_key(bool asyn
             async,
             ss => from g1 in ss.Set<Gear>()
                   from g2 in ss.Set<Gear>()
-                  // ReSharper disable once PossibleUnintendedReferenceComparison
+                      // ReSharper disable once PossibleUnintendedReferenceComparison
```
Comment (ErikEJ) [inline test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs:2801]: Undo unrelated changes

**Review by roji — APPROVED** [review]
@sulton-max I pushed a commit that removes the unrelated formatting changes and will merge as soon as the build is green - thanks for your contribution.

FYI it's important to review all your proposed changes before submitting a PR to make sure that nothing unintended slips in (like the above formatting stuff), a PR should ideally change exactly what it's supposed to, nothing more.

**Comment by sulton-max** [comment]
@roji Understood

@roji, @ErikEJ thank you guys so much for helping throughout the process

**Comment by ErikEJ** [comment]
@sulton-max You are welcome, congrats on your first contribution to EF Core

**Comment by sulton-max** [comment]
@ErikEJ @roji thank you once again, I'm really happy, would you mind if I ask help when working on other issues too ?

**Comment by roji** [comment]
@sulton-max we're always happy to receive contributions and to guide contributors where needed. But it's important for you to tackle tasks that you can overall reasonably accomplish by yourself, otherwise the guidance that we provide requires more time and effort than directly implementing the feature.
